docs/posts/querying-pypi-metadata-compiled-languages/index.qmd

```diff
@@ -0,0 +1,267 @@
+---
+title: Querying every file in every release on the Python Package Index (redux)
+author: Gil Forsyth
+date: 2023-11-15
+categories:
+    - blog
+---
+
+Seth Larson wrote a great [blog
+post](https://sethmlarson.dev/security-developer-in-residence-weekly-report-18)
+on querying a PyPI dataset to look for trends in the use of memory-safe
+languages in Python.
+
+You should check out Seth's article for more information on the dataset (and
+it's a good read!) but it caught our eye because it makes use of
+[DuckDB](https://duckdb.org/) to clean the data for analysis.
+
+That's right up our alley here in Ibis land, so let's see if we can duplicate
+Seth's results (and then continue on to plot them!)
+
+## Grab the data (locations)
+
+Seth showed (and then safely decomposed) a nested `curl` statement and that's
+always viable -- we're in Python land so why not grab the filenames using
+`urllib3`?
+
+```{python}
+import urllib3
+
+http = urllib3.PoolManager()
+
+resp = http.request("GET", "https://github.com/pypi-data/data/raw/main/links/dataset.txt")
+
+parquet_files = resp.data.decode().split()
+parquet_files
+```
+
+## Grab the data
+
+Now we're ready to get started with Ibis!
+
+DuckDB is clever enough to grab parquet metadata.  We can use `read_parquet` to
+create a lazy view of the parquet files and then build up our expression.
+
+```{python}
+import ibis
+from ibis import _  # <1>
+
+ibis.options.interactive = True
+```
+
+1. See ... for docs on the deferred operator!
+
+We can create a DuckDB connection:
+
+```{python}
+con = ibis.duckdb.connect()
+```
+
+And now let's load up one of the files (we can run the full query after)!
+
+```{python}
+pypi = con.read_parquet(parquet_files[0], table_name="pypi")
+```
+
+```{python}
+pypi.schema()
+```
+
+## Query crafting
+
+Let's break down what we're looking for. As a high-level view of the use of
+compiled languages, Seth is using file extensions as an indicator that a given
+filetype is used in a Python project.
+
+The dataset we're using has _every file in every project_ -- what criteria should we use?
+
+We can follow Seth's lead and look for things:
+
+1. A file extension that is one of: `asm`, `cc`, `cpp`, `cxx`, `h`, `hpp`, `rs`, `go`, and variants of `F90`, `f90`, etc...
+   That is, C, C++, Assembly, Rust, Go, and Fortran.
+2. We exclude matches where the file path is within the `site-packages/` directory.
+3. We should exclude matches that are in directories used for testing.
+
+```{python}
+expr = pypi.filter(
+    [
+        _.path.re_search(r"\.(asm|c|cc|cpp|cxx|h|hpp|rs|[Ff][0-9]{0-2}(?:or)?|go)$"),
+        ~_.path.re_search(r"(^|/)test(|s|ing)"),
+        ~_.path.contains("/site-packages/"),
+    ]
+)
+```
+
+```{python}
```

Comment:
```suggestion
```

```{python}
```

I always find the separate render visually odd, but up to you whether you want to leave it in.